File: 1/2021.06.28/139.Copy List with Random Pointer/python/solution.py
class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        if not head:
            return
        node = head
        while node:
            new_node = Node(node.val, node.next)
            node.next = new_node
            node = new_node.next
        # set random
        node = head
        while node:
            new_node = node.next
            new_node.random = None if node.random is None else node.random.next
            node = new_node.next
        # set next
        res = head.next
        node = head
        new_node = head.next
        while node:
            node.next = new_node.next
            new_node.next = None if new_node.next is None else new_node.next.next
            node = node.next
            new_node = new_node.next
        return res

File: 1/2021.06.28/139.Copy List with Random Pointer/java/solution.java
class Solution {

    private Map<Node, Node> visited = new HashMap<>();

    private Node getCloneNode(Node head) {
        if (head == null) {
            return null;
        }
        if (!visited.containsKey(head)) {
            visited.put(head, new Node(head.val));
        }
        return visited.get(head);
    }

    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }
        Node oldNode = head;
        Node newNode = new Node(head.val);
        visited.put(oldNode, newNode);
        while (oldNode != null) {
            newNode.random = getCloneNode(oldNode.random);
            newNode.next = getCloneNode(oldNode.next);
            oldNode = oldNode.next;
            newNode = newNode.next;
        }
        return visited.get(head);
    }
}
